website/docs/03-checker-bug-types.md

```diff
@@ -0,0 +1,1042 @@
+- [StrongSelf Not Checked](/checkers-bug-types#STRONG_SELF_NOT_CHECKED)
+- [Thread-safety violation](/checkers-bug-types#THREAD_SAFETY_VIOLATION)
+- [UI Thread Starvation](/checkers-bug-types#STARVATION)
+- [Unsafe_GuardedBy_Access](/checkers-bug-types#UNSAFE_GUARDEDBY_ACCESS)
+
+<a name="CHECKERS_IMMUTABLE_CAST"></a>
+
+## Checkers immutable cast
+
+This error type is reported in Java. It fires when an immutable collection is
+returned from a method whose type is mutable.
+
+```java
+  public List<String> getSomeList() {
+    ImmutableList<String> l = foo(...);
+    return l;
+  }
+```
+
+This can lead to a runtime error if users of `getSomeList` try to modify the
+list e.g. by adding elements.
+
+Action: you can change the return type to be immutable, or make a copy of the
+collection so that it can be modified.
+
+<a name="DEADLOCK"></a>
+
+## Deadlock
+
+This error is currently reported in Java. A deadlock occurs when two distinct
+threads try to acquire two locks in reverse orders. The following code
+illustrates a textbook example. Of course, in real deadlocks, the lock
+acquisitions may be separated by deeply nested call chains.
+
+```java
+  public void lockAThenB() {
+    synchronized(lockA) {
+      synchronized(lockB) {
+       // do something with both resources
+      }
+    }
+  }
+
+  public void lockBThenA() {
+    synchronized(lockB) {
+      synchronized(lockA) {
+       // do something with both resources
+      }
+    }
+  }
+```
+
+The standard solution to a deadlock is to fix an order of lock acquisition and
+adhere to that order in all cases. Another solution may be to shrink the
+critical sections (i.e., the code executing under lock) to the minimum required.
+
+Old-style containers such as `Vector` are synchronized on the object monitor,
+which means that deadlocks can occur even without explicit synchronisation on
+both threads. For instance:
+
+```java
+  public void lockAThenAddToVector() {
+    synchronized(lockA) {
+      vector.add(object);
+    }
+  }
+
+  public void lockVectorThenA() {
+    synchronized(vector) {
+      synchronized(lockA) {
+       // do something with both resources
+      }
+    }
+  }
+```
+
+Infer has support for detecting these deadlocks too.
+
+To suppress reports of deadlocks in a method `m()` use the
+`@SuppressLint("DEADLOCK")` annotation, as follows:
+
+```java
+  import android.annotation.SuppressLint;
+
+  @SuppressLint("DEADLOCK")
+  public void m() {
+  ...
+  }
+```
+
+<a name="DEAD_STORE"></a>
+
+## Dead store
+
+This error is reported in C++. It fires when the value assigned to a variables
+is never used (e.g., `int i = 1; i = 2; return i;`).
+
+<a name="EMPTY_VECTOR_ACCESS"></a>
+
+## Empty vector access
+
+This error type is reported only in C++, in versions >= C++11.
+
+The code is trying to access an element of a vector that Infer believes to be
+empty. Such an access will cause undefined behavior at runtime.
+
+```c++
```

Comment:
`c++` doesn't seem to be recognised any more: the code lost its syntax highlighting. `cpp` works. (fortunately there's only this one occurrence in the whole website!)

Reply:
Fixed